Data/Scripts/001_Technical/002_RubyUtilities.rb

```diff
@@ -162,6 +159,145 @@ def self.move(source, destination)
   end
 end
 
+#===============================================================================
+#  Extensions for the `Color` class
+#===============================================================================
+class Color
+  # alias for old constructor
+  alias init_original initialize unless self.private_method_defined?(:init_original)
+  #-----------------------------------------------------------------------------
+  # new constructor accepts RGB values as well as a hex number or string value
+  #-----------------------------------------------------------------------------
+  def initialize(*args)
+  	pbPrintException("Wrong number of arguments! At least 1 is needed!") if args.length < 1
+  	if args.length == 1
+      if args[0].is_a?(Fixnum)
+        hex = args[0].dup.to_s(16)
+      elsif args[0].is_a?(String)
+        hex = args[0].dup
+        hex.gsub!("#", "") if hex.include?("#")
+      end
+      pbPrintException("Wrong type of argument given!") if !hex
+      r = hex[0...2].to_i(16)
+      g = hex[2...4].to_i(16)
+      b = hex[4...6].to_i(16)
+  	elsif args.length == 3
+      r, g, b = *args
+  	end
+  	return init_original(r, g, b) if r && g && b
+  	return init_original(*args)
+  end
+  #-----------------------------------------------------------------------------
+  # returns an RGB color value as a hex value
+  #-----------------------------------------------------------------------------
+  def to_hex
+  	r = sprintf("%02X", self.red)
+  	g = sprintf("%02X", self.green)
+  	b = sprintf("%02X", self.blue)
+  	return ("#" + r + g + b).upcase
+  end
+  #-----------------------------------------------------------------------------
+  # returns Hex color value as RGB
+  #-----------------------------------------------------------------------------
+  def to_rgb(hex)
+  	hex = hex.to_s(16) if hex.is_a?(Numeric)
+  	r = hex[0...2].to_i(16)
+  	g = hex[2...4].to_i(16)
+  	b = hex[4...6].to_i(16)
+  	return r, g, b
+  end
+  #-----------------------------------------------------------------------------
+  # returns decimal color
+  #-----------------------------------------------------------------------------
+  def to_dec
+  	return self.to_hex.delete("#").to_i(16)
+  end
+  #-----------------------------------------------------------------------------
+  # byte order in ARGB instead of RGBA.
+  #-----------------------------------------------------------------------------
+  def to_i
+    return self.alpha.to_i << 24 | self.blue.to_i << 16 | self.green.to_i << 8 | self.red.to_i
```

Comment:
One of this and `def self.from_i` has blue and red the wrong way round. I'm not sure which.

Reply:
It should be `ABGR` from my understanding; but I had some super strange things happening in MKXP in the `self.from_i`. For some reason the `BGRA` is the correct way to work it there. I'm not exactly sure why when going from int -> RGBA it works like that ... but I've tested this in EBDX extensively.

Reply:
Isn't `def self.from_i` as written in the order of ARGB? Contrasting with `def to_i` which is ABGR. That implies that if you do `Color.from_i(Color.red.to_i)`, you'll end up with blue rather than red. The comment description for `def to_i` says it's in ARGB format, when it's clearly returning a number in ABGR.

It's bizarre that `to_i` and `from_i` put the numbers in different orders. In fact, since we're the ones writing these methods, they should work the same. Surely mkxp-z has nothing to do with this?

Reply:
The comment above is definitely does not match up with what's in the code - I'll test it again on proper values from unpacking a bitmap and in some additional test case to see how it behaves again.

Reply:
Have you had a chance to look at this yet?

Reply:
Sorry, not in any detail yet. Been a bit busy with work. However, with some quick testing -> you're right: going from color to int back to color shifts the order of the `red` value. There's a discrepancy with the way the raw unpacked bitmap data is calculated though, and the int values from a normal `Color` object do not align what's in the unpacked array. I'll need to dig deeper into this.